pkg/sql/ir/ir.go

```diff
@@ -76,17 +76,34 @@ type Executor interface {
 type SQLStatement interface {
 	SetOriginalSQL(string)
 	Execute(Executor) error
+	IsExtended() bool
+	GetOriginalSQL() string
 }
 
-// TrainStmt is the intermediate representation for code generation of a training job.
-type TrainStmt struct {
+// ExtendedSQL contains common fields of TrainStmt, PredictStmt and AnalyzeStmt
+type ExtendedSQL struct {
+	// Select specifies the query for fetching the training data. For example, "select * from iris.train;".
+	Select string
+	// DataSource contains the connection information. For example, "hive://root:root@localhost:10000/churn"
+	DataSource string
+	// Attributes is a map of parsed attribute in the WITH Clause. For example, after parsing
+	// "select ... train ... with train.epoch = 1000, model.hidden_units = [10, 10]",
+	// the Attributes will be {"train.epoch": 1000, "model.hidden_units": [10 10]}.
+	Attributes map[string]interface{}
 	// OriginalSQL record the original SQL statement used to get current IR result
 	// FIXME(typhoonzero): OriginalSQL is a temporary field. Can remove this when all moved to IR
 	OriginalSQL string
-	// DataSource contains the connection information. For example, "hive://root:root@localhost:10000/churn"
-	DataSource string
-	// Select specifies the query for fetching the training data. For example, "select * from iris.train;".
-	Select string
+}
+
+// IsExtended returns whether a SQLStatement is an extended SQL statement
+func (sql *ExtendedSQL) IsExtended() bool { return true }
+
+// GetOriginalSQL returns the original SQL statement used to get current IR result
+func (sql *ExtendedSQL) GetOriginalSQL() string { return sql.OriginalSQL }
+
+// TrainStmt is the intermediate representation for code generation of a training job.
+type TrainStmt struct {
+	ExtendedSQL
```

Comment:
`TrainStmt. ExtendedSQL` seems confusing. Can direct implement `GetOriginalSQL` and `IsExtended` for each `TrainStmt`, `PredictStmt` and `AnalyzeStmt`

Reply:
> `TrainStmt. ExtendedSQL` seems confusing. 

Structure embedding in Go has similar semantics as private inheritance in C++: they both disguise composition as inheritance. So we never need to write `TrainStmt.ExtendedSQL` or `TrainStmt::ExtendedSQL` explicitly. Struct embedding is good for extracting common fields/methods of several similar structs into a single struct.


> Can direct implement `GetOriginalSQL` and `IsExtended` for each `TrainStmt`, `PredictStmt` and `AnalyzeStmt`

I don't think code duplication is a good idea, maybe it's the naming of `ExtendedSQL` that causes the discomfort, how about renaming it as `ExtendedSQLBase`?  i.e. We have one `StandardSQL` and three `ExtendedSQL`s, the latter share implementation from `ExtendedSQLBase`.